### PR TITLE
Native Philips Hue Suport

### PR DESCRIPTION
**Pull request by wolfcomp**
Closes: #1647 

### Adds
Philips Hue syncing

### Known issues that are impossible to fix in Aurora
If you have a event set in the hue app on the phone it will try to activate that event on the lights but Aurora takes over the light syncing again the next time it sends to the lights making a slight blink effect.

### Images
931961e300
a215428906

Note all the previous config will work with this new version as this is only for splitting the pull requests. And to remove my custom funky mess when it comes to how i have to build a custom version for it to work fully on my setup.

### CODE REVIEW

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs [hunk 1]
```diff
@@ -97,116 +94,36 @@ public void SetVariable(object newvalue)
...
-                    Value = b.Value;
-                    break;
-                case BindableColor b:
-                    Value = b.Value;
-                    break;
-                case Bindable<string> b:
-                    Value = b.Value;
-                    break;
+                Value = (Bindable as Bindable<object>).Value;
+                (Bindable as Bindable<object>).ValueChanged += _ => Value = (Bindable as Bindable<object>).Value;
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: As before, you cannot cast a generic type to a different generic type even if the type parameter of one extends the type param of the other.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs [hunk 2]
```diff
@@ -1,5 +1,7 @@
 ﻿using Aurora.Utils;
+
 using Newtonsoft.Json;
+
 using System;
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: What's with these sort of unnecessary changes? We've not got any whitespace in the imports in other files.

File: Project-Aurora/Project-Aurora/Devices/DeviceManager.cs [hunk 3]
```diff
@@ -96,23 +96,24 @@ public int RetryAttempts
...
+            devices.Add(new DeviceContainer(new Devices.UnifiedHID.UnifiedHIDDevice()));     // UnifiedHID
+            devices.Add(new DeviceContainer(new Devices.Wooting.WootingDevice()));           // Wooting
+            devices.Add(new DeviceContainer(new Devices.Creative.SoundBlasterXDevice()));    // SoundBlasterX
+            devices.Add(new DeviceContainer(new Devices.LightFX.LightFxDevice()));           // Alienware
+            devices.Add(new DeviceContainer(new Devices.Dualshock.DualshockDevice()));       // DualShock 4
+            devices.Add(new DeviceContainer(new Devices.Drevo.DrevoDevice()));               // Drevo
+            devices.Add(new DeviceContainer(new Devices.YeeLight.YeeLightDevice()));         // YeeLight
+            devices.Add(new DeviceContainer(new Devices.Asus.AsusDevice()));                 // Asus
+            devices.Add(new DeviceContainer(new Devices.NZXT.NZXTDevice()));                 // NZXT
+            devices.Add(new DeviceContainer(new Devices.Hue.HueDevice()));                   // Philips Hue
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Devices/DeviceManager.cs]: There's really no need for these sorts of very minor cosmetic changes, especially when it falls outside of the scope of your PR.

File: Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+using Aurora.Utils;
+
+using Corale.Colore.Annotations;
+
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Aurora.Settings.Bindables
+{
+    public class Bindable<T> : IBindable<T>, ISerializableBindable
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs]: What sort of issues are these "Bindables" solving that aren't already solved by the existing VarRegistry. While the VarRegistry is by no stretch of the imagination a great example of config management, it has been doing the job for other devices for a long time. Whilst we are not against changing the VarRegistry by any means, having spoken to the other contributors, we currently do not see any advantage of using this over the existing method and we certainly don't want to have multiple config management systems doing effectively the same thing.

File: Project-Aurora/Project-Aurora/Settings/Bindables/BindableNumber.cs [hunk 5]
```diff
@@ -0,0 +1,428 @@
...
+                    break;
+
+                case TypeCode.Double:
+                    var doubleBindable = this as BindableNumber<double>;
+                    if (doubleBindable == null) throw new ArgumentNullException(nameof(doubleBindable), $"Generic type {typeof(T)} does not match actual bindable type {GetType()}.");
+
+                    doubleBindable.Value += Convert.ToDouble(val);
+                    break;
+            }
+        }
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/Bindables/BindableNumber.cs]: Is there any reason to use switch trees like this? Why not get the type of `this` and then from there get the generic parameter that has been defined on it and use `Convert.ChangeType` and pass the generic parameter type in? It would mean you could remove a lot of these big chunks of code.

File: Project-Aurora/Project-Aurora/Settings/Bindables/ExtraBindables.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+                {
+                    var k = f.Split('=');
+                    return new KeyValuePair<string, string>(k[0], k[1]);
+                }).ToDictionary(f => f.Key, f => f.Value);
+                return new RealColor(Color.FromArgb(Convert.ToInt32(h["A"]), Convert.ToInt32(h["R"]), Convert.ToInt32(h["G"]), Convert.ToInt32(h["B"])));
+            }
+            set => base.Default = value.ToString();
+        }
+    }
+}
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/Bindables/ExtraBindables.cs]: Why are you storing this as a string? It means you're having to parse and stringify the value whenever it's read and written.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs [hunk 7]
```diff
@@ -20,21 +24,66 @@ public class VariableRegistryItem
...
+                    break;
+                case Bindable<string> b:
+                    Value = b.Value;
+                    Default = b.Default;
+                    break;
+            }
+            Title = title;
+            Remark = remark;
+            Flags = flags;
         }
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: What if I want to add a new bindable type? Do I have to add it in here too? This has the potential to become another massive chunk of code. And what about the other BindableNumbers you have, such as short?

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs [hunk 8]
```diff
@@ -231,13 +288,12 @@ public void RemoveVariable(string name)
...
+            string str = JsonConvert.SerializeObject(this, Formatting.None, new JsonSerializerSettings {TypeNameHandling = TypeNameHandling.All, Binder = Aurora.Utils.JSONUtils.SerializationBinder});
             return JsonConvert.DeserializeObject(
-                    str,
-                    this.GetType(),
-                    new JsonSerializerSettings { ObjectCreationHandling = ObjectCreationHandling.Replace, TypeNameHandling = TypeNameHandling.All, Binder = Aurora.Utils.JSONUtils.SerializationBinder }
-                    );
+                str,
+                this.GetType(),
+                new JsonSerializerSettings {ObjectCreationHandling = ObjectCreationHandling.Replace, TypeNameHandling = TypeNameHandling.All, Binder = Aurora.Utils.JSONUtils.SerializationBinder}
+            );
```
Comment (Wibble199) [inline Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: You really need to stop committing code that changes styling changes like this. It's unnecessary, and the project's general style is to have spaces around the braces. Even if your IDE is doing these changes automatically, you should be looking at what you are committing before doing so intead of just adding all changes.

**Review by Wibble199 — COMMENTED** [review]
Not looked over all the code but these are some of the things that stood out at problematic.

File: Project-Aurora/Project-Aurora/Settings/Bindables/ExtraBindables.cs (hunk 6 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/Bindables/ExtraBindables.cs]: The reason comes down to how the color is stored in run time where i need to export it reliably to the config witch is a JSON in a seperate file and how the config is loaded as this extends a Bindable<string> and not a color definition.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 8 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: Honestly you guys should have looked at supplying a ReSharper settings file that i could use so i don't have these issues in the repo.

File: Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs (hunk 4 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs]: When i wrote the hue code and i tried to actualy get the values back from the VarRegistry they were the default values and never returned the actual changed values and when i implemented this it fixed the issue.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 2 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: https://github.com/antonpup/Aurora/pull/1757#discussion_r336581440

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs [hunk 9]
```diff
@@ -97,116 +94,36 @@ public void SetVariable(object newvalue)
...
-                            b.Value = (long)newvalue;
-                            break;
-                        case BindableColor b:
-                            b.Value = (RealColor)newvalue;
-                            break;
-                        case Bindable<string> b:
-                            b.Value = (string)newvalue;
-                            break;
-                    }
+                    (Bindable as Bindable<object>).Value = newvalue;
```
[the remark this answers is not in the document]
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: Yes i tested this and all the settings was being saved and stored properly in the code.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 1 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: https://github.com/antonpup/Aurora/pull/1757#discussion_r336582007

File: Project-Aurora/Project-Aurora/Devices/DeviceManager.cs (hunk 3 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Devices/DeviceManager.cs]: Honestly those comments don't need to be there _as mentioned in the discord_ since we can clearly see what is what

File: Project-Aurora/Project-Aurora/Settings/Bindables/BindableNumber.cs (hunk 5 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/Bindables/BindableNumber.cs]: I would need to test that out before i can comment further than this is a solution i found from another repo that is MIT licenced.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 7 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: A tiny little miss yesterday when i changed most of the switches over to Bindable<object> as i was really tired when i made that change.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Wibble199) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: But why chose to create a whole new system over attempting to resolve the issue in the existing code?

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 8 above)
Comment (Wibble199) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: I don't think it's really on us to have to provide a file for a 3rd party proprietry software.

File: Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs (hunk 4 above)
Comment (Wibble199) [reply Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs]: But why did you choose to implement a new system just for the Hue over attempting to fix the code that is already in place in many other places?

Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/Bindables/Bindable.cs]: As when i break pointed the code it wasn't even changed in the VarRegistry meaning it was something in WPF that most likely caused it not to be set.

File: Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs (hunk 8 above)
Comment (wolfcomp) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: ReSharper isn't a proprietry software and is actual being used quite the bit in actual companies. There exists a 30 day trail for it meaning it can be configured up to how the code should be removing these issues later down the line.

Comment (Wibble199) [reply Project-Aurora/Project-Aurora/Settings/VariableRegistry.cs]: Right, but if you wanna keep using it beyond the 30 days you need to pay for it. Therefore, proprietry.

Either way, if it is or isn't on us to provide a config file, it doesn't mean it's open to random style changes that add nothing.

**Comment by RaberShef** [comment]
This merge is waiting since last year can someone merge it already?

**Comment by wtf911** [comment]
When is this going to be added? Would love this to work with my Hue lights please!!!!!!!!!!!!!!

**Comment by wolfcomp** [comment]
Will close this for now as im to busy to keep up with the code changes and other stuff going on.
